marker_api/src/ast/common/span.rs

```diff
@@ -2,133 +2,469 @@ use std::marker::PhantomData;
+/// Check out the documentation of [`ExpnInfo`].
 #[repr(C)]
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Span<'ast> {
-    source: &'ast SpanSource<'ast>,
-    /// The start marks the first byte in the [`SpanSource`] that is included in this
-    /// span. The span continues until the end position.
-    start: usize,
-    end: usize,
+    _lifetime: PhantomData<&'ast ()>,
+    /// The source of this [`Span`]. The id space and position distribution is
+    /// decided by the driver. To get the full source information it might be
+    /// necessary to also pass the start and end position to the driver.
+    source_id: SpanSrcId,
+    /// This information could also be retrieved, by requesting the [`ExpnInfo`]
+    /// of this span. However, from looking at Clippy and rustc lints, it looks
+    /// like the main interested is, if this comes from a macro expansion, not from
+    /// which one. Having this boolean flag will be sufficient to answer this simple
+    /// question and will save on extra [`SpanSrcId`] mappings.
+    from_expansion: bool,
+    start: SpanPos,
+    end: SpanPos,
+}
+
+impl<'ast> std::fmt::Debug for Span<'ast> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fn fmt_pos(pos: Option<FilePos<'_>>) -> String {
+            match pos {
+                Some(pos) => format!("{}:{}", pos.line(), pos.column()),
+                None => "[invalid]".to_string(),
+            }
+        }
+
+        let src = self.source();
+        let name = match src {
+            SpanSource::File(file) => format!(
+                "{}:{} - {}",
+                file.file(),
+                fmt_pos(file.to_file_pos(self.start)),
+                fmt_pos(file.to_file_pos(self.end))
+            ),
+            SpanSource::Macro(expn) => format!("[Inside Macro] {:#?}", expn.call_site()),
+        };
+        f.debug_struct(&name).finish()
+    }
 }
 
 impl<'ast> Span<'ast> {
-    pub fn is_from_file(&self) -> bool {
-        matches!(self.source, SpanSource::File(..) | SpanSource::Sugar(..))
+    /// Returns `true`, if this [`Span`] comes from a macro expansion.
+    pub fn is_from_expansion(&self) -> bool {
+        self.from_expansion
+    }
+
+    /// Returns the code snippet that this [`Span`] refers to or [`None`] if the
+    /// snippet is unavailable.
+    ///
+    /// ```ignore
+    /// let variable = 15_000;
+    /// //             ^^^^^^
+    /// //             lit_span
+    ///
+    /// lit_span.snippet(); // -> Some("15_000")
+    /// ```
+    ///
+    /// There are several reasons, why a snippet might be unavailable. Also
+    /// depend on the used driver. You can also checkout the other snippet
+    /// methods to better deal with these cases:
+    /// * [`snippet_or`](Self::snippet_or)
+    /// * [`snippet_with_applicability`](Self::snippet_with_applicability)
+    #[must_use]
+    pub fn snippet(&self) -> Option<&'ast str> {
+        with_cx(self, |cx| cx.span_snipped(self))
```

Comment:
`[off-topic]`
By reading the `rustc` source code I happened to find the crate [`scoped-tls`](https://docs.rs/scoped-tls/latest/scoped_tls/) created by rustc devs. 

Maybe it could be used as backend for storing the AST context? It would remove some `unsafe` code from the guts of `with_cx`

Reply:
I don't really see how the linked crate would improve this much.

The main problem I see with the current solution is the lifetime transmutes, but I don't see a way to store `&'ast AstContext<'ast>` in a static variable otherwise.

Reply:
I think `scoped-tls` allows you to get rid of `&'ast` part, but it won't remove the `<'ast>` from the type, so you are right, it won't fix the transmutes.